### DISCUSSION TITLE
Breaking: Subcommands

### DISCUSSION THREAD

**Discussion by rawkode** [comtrya/comtrya#160] [Ideas]
At the moment, `./comtrya` only performs a single command: run manifests.

We accept a few flags that can change this behaviour, but there's no concept of a subcommand.

Before we move towards 1.0, we should definitely have the discussion around this behaviour.

Do we see a world with:

```shell
# Version
comtrya version

# Edit Comtrya.yaml
comtrya contexts --set name "David Flanagan"

# View all Contexts
comtrya contexts --dump -o yaml

# Import Local Dotfile
comtrya import file ~/.zsh/prompt.zsh zsh.prompt # to ./zsh/files/prompt.zsh

# Import AND Install a Package using the default provider
comtrya import package curl devtools.web # to ./devtools/web.yaml (or ./devtools/web/main.yaml if it exists)
comtrya import package curl devtools # to ./devtools.yaml (or ./devtools/main.yaml if it exists)
```

**Comment by rawkode** [comment]
Would love some feedback from @icepuma @kraileth @bketelsen @nunix @sidpalas @svallory @mt-inside @Spazzy757 @vielmetti @martintc

**Comment by icepuma** [comment]
When moving towards subcommands I recommend using `clap`.

Every subcommand to inspect the environment of `comtrya` comes in handy e.g. dumping all contexts or listing all plugins!

**Comment by bketelsen** [comment]
yes, most of those look very useful. I think subcommands are the only way to expand the functionality without getting caught up in a sea of flags. Ick.

**Comment by vielmetti** [comment]
I'd prefer an approach with commands to one that had lots of flags, all things considered. Especially if the flags all support some kind of "help" subcommand.

**Comment by rawkode** [comment]
OK. I'll get subcommand support added over the next two weeks. Thanks for your feedback, everyone.

**Reply by nunix** [reply]
really nice addition, and concerning subcommands, as I saw it for `k3d` "break" on v4, as an advice is to already think at the "verb-noun" VS "noun-verb" syntax.

This would avoid some breaking changes later on

**Comment by Lockszmith-GH** [comment]
+1 for subcommands, some that are really required:
* <blank> - show the help (following git, docker, gcloud, terraform, apt, etc...)
* completion <shell-name> - generate auto-completion code based on the shell (I know there is a go library for that, don't know if rust has one)

**Comment by svallory** [comment]
@rawkode 2 notes on the list above:

1. I think the syntax needs to be normalized across subcommands. If we have `comtrya import file ...` (my preferred syntax) we should have `comtrya contexts set` and not `comtrya contexts --set`

2. We are missing `comtrya contexts get`

**Reply by rawkode** [reply]
Good call
